### PR TITLE
Ensure that we don't root the compilation when using a custom exception marshalling type.

### PR DESCRIPTION
**Pull request by jkoritzinsky**
This is similar to #79051 but for the new VirtualMethodIndex generator.

### CODE REVIEW

File: src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/VirtualMethodIndexData.cs [hunk 1]
```diff
@@ -6,9 +6,37 @@
...
-    internal sealed record VirtualMethodIndexData(int Index) : InteropAttributeCompilationData
+    internal sealed record VirtualMethodIndexData(int Index) : InteropAttributeData
+    {
+        public bool ImplicitThisParameter { get; init; }
+
+        public MarshalDirection Direction { get; init; }
+
+        public bool ExceptionMarshallingDefined { get; init; }
+
+        public ExceptionMarshalling ExceptionMarshalling { get; init; }
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/VirtualMethodIndexData.cs:20]: Can't we just define all these in the ctor for the type? We are retyping things a lot below when it seems like we need all of them always.

File: src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/VirtualMethodIndexData.cs [hunk 2]
```diff
@@ -6,9 +6,37 @@
 namespace Microsoft.Interop
 {
     /// <summary>
-    /// VirtualMethodIndexAttribute data
+    /// Contains the data related to a VirtualMethodIndexAttribute, without references to Roslyn symbols.
+    /// See <seealso cref="VirtualMethodIndexCompilationData"/> for a type with a reference to the StringMarshallingCustomType
     /// </summary>
-    internal sealed record VirtualMethodIndexData(int Index) : InteropAttributeCompilationData
+    internal sealed record VirtualMethodIndexData(int Index) : InteropAttributeData
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/VirtualMethodIndexData.cs:12]: The difference between this and the Compilation version is a single `INamedTypeSymbol` field. Is this pattern what Roslyn recommends? It seems very unfortunate to have two types that are identical except for a field with a special "type". The comments are very helpful, but the subtle nature of how to use these types is a bit less than ideal.

Comment (jkoritzinsky) [reply src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/VirtualMethodIndexData.cs:12]: There's no well-defined pattern for this scenario. This design is following the current best practice of "Provide a data model object that doesn't reference any symbols or anything that would keep the Compilation object alive" (which is the best guidance we have today).

I'm open to alternative ideas/designs that would make this cleaner.

Comment (AaronRobinsonMSFT) [reply src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/VirtualMethodIndexData.cs]: Nope, if that is the pattern people are going with sounds good.

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Review by elinor-fung — APPROVED** [review]
We don't currently have `ComInterfaceGenerator` tests that check incremental generation, right?

**Comment by jkoritzinsky** [comment]
Not yet. We should be able to refactor our incremental tests from LibraryImportGenerator to be reusable for different generators.

**Pull request merged**